aptos-move/move-examples/ring_deque/sources/ring_deque.move

```diff
@@ -0,0 +1,296 @@
+/// Ring buffer implementation of a double-ended queue.
+///
+/// Has O(1) push/pop operations at both front and back, unlike a vector
+/// which has O(n) push/pop operations at the front.
+module ring_deque::ring_deque {
+
+    use std::option::{Self, Option};
+    use std::vector;
+
+    /// Buffer capacity must be nonzero.
+    const E_CAPACITY_ZERO: u64 = 0;
+    /// Ring deque is full;
+    const E_FULL: u64 = 1;
+    /// Ring deque is empty;
+    const E_EMPTY: u64 = 2;
+
+    /// Ring buffer implementation of a double-ended queue. Mechanics:
+    ///
+    /// 1. Uses a single vector of options to store elements, instantiated to full capacity.
+    /// 2. Options are used to extract elements, but not resize the vector.
+    /// 3. Bounds are maintained by front and back indices, which wrap around.
+    struct RingDeque<T> has copy, drop, store {
+        data: vector<Option<T>>,
+        capacity: u64,
+        front: u64,
+        back: u64,
+        length: u64
```

Comment:
Can't this be derived from back and front?

Reply:
@lightmark while it is algorithmically possible to derive `length = f(capacity, front, back)`, the `length` field is essentially a caching mechanism to improve performance and lookup time in context like, for example `public fun length`